coap-core/src/main/java/com/mbed/coap/server/internal/PutOnlyMapImpl.java

```diff
@@ -0,0 +1,123 @@
+    private long nextWarnMessage;
+    private final long duplicateDetectionTimeMillis;
+    private final String cacheName;
+    private final ScheduledExecutorService scheduledExecutor;
+    private ScheduledFuture<?> cleanWorkerFut;
+
+    public PutOnlyMapImpl(String cacheName,
+            long maxSize,
+            long duplicateDetectionTimeMillis,
+            long cleanIntervalMillis,
+            long warnIntervalMillis,
+            ScheduledExecutorService scheduledExecutor) {
+        this.cacheName = cacheName;
+        this.duplicateDetectionTimeMillis = duplicateDetectionTimeMillis;
+        this.maxSize = maxSize;
+        this.cleanIntervalMillis = cleanIntervalMillis;
+        this.warnIntervalMillis = warnIntervalMillis;
+        this.scheduledExecutor = scheduledExecutor;
+        this.overSizeMargin = maxSize / 100; //1%
+        underlying = new ConcurrentHashMap<>();
+    }
+
+    public void start() {
+        cleanWorkerFut = scheduledExecutor.scheduleWithFixedDelay(() -> clean(), cleanIntervalMillis, cleanIntervalMillis, TimeUnit.MILLISECONDS);
+    }
+
+    public void stop() {
+        cleanWorkerFut.cancel(true);
+    }
+
+    @Override
+    public V putIfAbsent(K key, V value) {
+        V result = underlying.putIfAbsent(key, value);
+        // Cleanup only if new entry was added to map.
+        if (result == null) {
+            cleanupBulk();
+        }
+        return result;
+
+    }
+
+    @Override
+    public void put(K key, V value) {
+        underlying.put(key, value);
+    }
+
+    public void clean() {
+        int removedItems = 0;
+        Iterator<K> it = underlying.keySet().iterator();
+        final long currentTimeMillis = System.currentTimeMillis();
+        while (it.hasNext()) {
+            if (currentTimeMillis - it.next().getCreatedTimestampMillis() > duplicateDetectionTimeMillis) {
+                it.remove();
+                removedItems++;
+            }
+        }
+        if (LOGGER.isTraceEnabled() && removedItems > 0) {
+            LOGGER.trace("CoAP request duplicate list, non valid items removed: " + removedItems + " ");
+        }
+    }
+
+    public void cleanupBulk() {
+        if (underlying.size() > maxSize + overSizeMargin && REDUCE_LOCK.tryLock()) {
+            try {
+                //reduce map size in bulk
+                Iterator<K> it = underlying.keySet().iterator();
+                try {
+                    for (int i = 0; i <= overSizeMargin && it.hasNext(); i++) {
+                        it.next();
+                        it.remove();
+                        //requestList.remove(it.next());
+                    }
+                } catch (Exception e) {
+                    LOGGER.error(e.getMessage(), e);
+                }
+
+                if (nextWarnMessage < System.currentTimeMillis()) {
+                    LOGGER.warn(cacheName + " has reached max size (" + maxSize + "), reduced by " + overSizeMargin);
```

Comment:
Does this cache name really be passed in? To me hard coded log would be just fine here.

Reply:
I think it would help a lot. This way we'd know which cache is full.